File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Main {
    private static final int INF = (int) 1e9;
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        int n = nextInt();
        int m = nextInt();

        st = new StringTokenizer(br.readLine());
        int start = nextInt();

        List<Node>[] edges = new List[n + 1];

        for (int i = 1; i <= n; i++) {
            edges[i] = new ArrayList<>();
        }

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = nextInt();
            int v = nextInt();
            int w = nextInt();

            edges[u].add(new Node(v, w));
        }

        PriorityQueue<Node> pq = new PriorityQueue<>(n, Comparator.comparingInt(a -> a.w));
        int[] dist = new int[n + 1];

        pq.add(new Node(start, 0));
        Arrays.fill(dist, INF);
        dist[start] = 0;

        while (!pq.isEmpty()) {
            Node u = pq.poll();

            if (u.w != dist[u.v]) continue;

            for (Node edge : edges[u.v]) {
                int dw = u.w + edge.w;

                if (dw < dist[edge.v]) {
                    dist[edge.v] = dw;
                    pq.add(new Node(edge.v, dw));
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int u = 1; u <= n; u++) {
            sb.append(dist[u] == INF ? "INF" : dist[u]).append('\n');
        }

        System.out.println(sb);
        br.close();
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }

    static class Node {
        int v;
        int w;

        Node(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }
}

File: 백준/Gold/1753. 최단경로/최단경로.kt
import java.io.StreamTokenizer
import java.util.*

const val INF = 1e9.toInt()

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val readInt = {
        nextToken()
        nval.toInt()
    }
    val n = readInt()
    val m = readInt()
    val root = readInt() - 1
    val edges = Array(n) { mutableMapOf<Int, Int>() }

    repeat(m) {
        val u = readInt() - 1
        val v = readInt() - 1
        val w = readInt()

        if ((edges[u][v] ?: 11) > w) edges[u][v] = w
    }

    val dist = IntArray(n) { INF }
    val pq = PriorityQueue<Int>(compareBy { dist[it] })

    dist[root] = 0
    pq.add(root)

    while (pq.isNotEmpty()) {
        val u = pq.poll()

        for ((v, w) in edges[u]) {
            val d = dist[u] + w

            if (dist[v] > d) {
                dist[v] = d
                pq.add(v)
            }
        }
    }

    with(System.out.bufferedWriter()) {
        dist.forEach { append("${if (it == INF) "INF" else it}\n") }
        flush()
        close()
    }
}